File: projeto-haskell/src/Controller/QuizController.hs
{-# LANGUAGE OverloadedStrings #-}
module Controller.QuizController where

import Entities.Quiz
import Database.SQLite.Simple
import Database.SQLite.Simple.FromRow
import Database.SQLite.Simple.ToField
import Utils.Util

data QuizAnswer = QuizAnswer {
  quizId:: String,
  name:: String,
  topic:: String,
  userAnswerId:: String,
  score:: Double,
  rating:: Int,
  created_at:: String
}

getQuizAnswerId:: QuizAnswer -> String
getQuizAnswerId = quizId
instance Show QuizAnswer where
    show (QuizAnswer quizId name topic user_answer_id score rating created_at) =
        "Quiz: "++name++", Tópico: "++topic++", "++
        "Pontuação: "++formatFloatN score 2++", Avaliação: "++show rating++
        ", Respondeu em: "++created_at
instance FromRow Quiz where
  fromRow = Quiz <$> field
                 <*> field
                 <*> field
                 <*> field
                 <*> field

instance FromRow QuizAnswer where
  fromRow = QuizAnswer <$> field
                       <*> field
                       <*> field
                       <*> field
                       <*> field
                       <*> field
                       <*> field
instance ToRow Quiz  where
  toRow (Quiz id name topic user_id created_at) = toRow (id, name, topic, user_id)                      

addQuiz :: String -> String -> String -> IO String
addQuiz nameQuiz topicQuiz userId = do
    conn <- open dbPath
    uuidQuiz <- getRandomUUID
    let quiz = Quiz uuidQuiz nameQuiz topicQuiz userId ""
    execute conn "INSERT INTO quiz (quiz_id,name,topic,user_id,created_at)\
    \ VALUES (?,?,?,?,datetime('now','localtime'))" quiz
    return uuidQuiz
      
getMyQuizzes :: String -> IO [Quiz]
getMyQuizzes user_id = do
    conn <- open dbPath
    query conn "SELECT * from quiz WHERE user_id = ?" (Only user_id) :: IO [Quiz]

getAllQuizzes :: IO [Quiz]
getAllQuizzes = do
    conn <- open dbPath
    query_ conn "SELECT * from quiz" :: IO [Quiz]

getAllQuizzesWithQuestions :: IO [Quiz]
getAllQuizzesWithQuestions = do
    conn <- open dbPath
    query_ conn "SELECT DISTINCT qz.quiz_id, qz.name, qz.topic, qz.user_id,qz.created_at\
    \ FROM quiz qz, question qe WHERE qz.quiz_id = qe.quiz_id" :: IO [Quiz]

getAllQuizzesWithAnswers :: String -> IO [Quiz]
getAllQuizzesWithAnswers user_id = do
    conn <- open dbPath
    query conn "SELECT DISTINCT qz.quiz_id, qz.name, qz.topic, qz.user_id,qz.created_at\
      \ FROM quiz qz, user_answer ua WHERE qz.quiz_id = ua.quiz_id\
      \ AND ua.user_id=?" (Only user_id) :: IO [Quiz]
    
getAllQuizzesAnswers :: String -> IO [QuizAnswer]
getAllQuizzesAnswers user_id = do
    conn <- open dbPath
    query conn "SELECT q.quiz_id, q.name, q.topic, ua.user_answer_id, \
    \ua.score, ua.rating, ua.created_at FROM quiz q, user_answer ua \
    \WHERE q.quiz_id = ua.quiz_id AND ua.user_id = ?" (Only user_id) :: IO [QuizAnswer]

-- retorna todos os quizzes que o usuário respondeu, porém sem repetir o quiz
getAllQuizzesAnswersUnique :: String -> IO [Quiz]
getAllQuizzesAnswersUnique user_id = do
    conn <- open dbPath
    query conn "SELECT DISTINCT q.quiz_id,q.name,q.topic,q.user_id,q.created_at \
      \FROM quiz q, user_answer ua WHERE q.quiz_id = ua.quiz_id AND \
      \ua.user_id = ?" (Only user_id) :: IO [Quiz]

updateQuiz:: Quiz -> IO()
updateQuiz quiz = do
      conn <- open dbPath
      execute conn "UPDATE quiz SET name = ?, topic = ? WHERE quiz_id = ?"
        (getName quiz :: String,getTopic quiz, getIdQuiz quiz :: String)

deleteQuiz :: String -> IO()
deleteQuiz quizId = do
    conn <- open dbPath
    execute_ conn "PRAGMA foreign_keys = ON"
    execute conn "DELETE FROM quiz WHERE quiz_id = ?" (Only quizId)

File: projeto-prolog/controllers/quizController.pl
:- module(quizController, 
    [createQuiz/4,
     getAllQuizzes/1,
     printQuizzes/2,
     getAllMyQuizzes/2,
     deleteQuiz/1,
     printQuiz/1,
     updateQuiz/3,
     getAllQuizzesWithQuestions/1,
     getAllUserAnsweredQuizzes/2,
     getAllUserAnsweredQuizzesUnique/2,
     getAllQuizzesAnswers/2,
     printQuizAnswers/2
    ]).

:- use_module( library(prosqlite) ).

% cria um quiz a partir do nome do quiz, tópico do quiz e id do usuário
createQuiz(Name, Topic, UserId, UUIDQuiz) :-
    uuid(QuizUUID),
    format(atom(Query), "INSERT INTO quiz (quiz_id,name,topic,user_id,created_at) VALUES ('~w','~w','~w','~w',datetime('now','localtime'));",
        [QuizUUID,Name,Topic,UserId]),
    sqlite_query( db, Query, _), UUIDQuiz = QuizUUID.

% formato do retorno: row(QuizId,Name,Topic,UserId,CreatedAt)
getAllQuizzes(Quizzes) :-
    format(atom(Query), "SELECT * FROM quiz;",[]),
    findall( Row, sqlite_query(db, Query, Row), Quizzes ).

% formato do retorno: [row(QuizId,Name,Topic,UserId,CreatedAt)]
getAllMyQuizzes(UserId,Quizzes) :-
    format(atom(Query), "SELECT * FROM quiz WHERE user_id = '~w';",[UserId]),
    findall( Row, sqlite_query(db, Query, Row), Quizzes ).

getAllQuizzesWithQuestions(Quizzes) :-
    format(atom(Query), "SELECT DISTINCT qz.quiz_id, qz.name, qz.topic, qz.user_id,qz.created_at 
        FROM quiz qz, question qe WHERE qz.quiz_id = qe.quiz_id;",[]),
    findall( Row, sqlite_query(db, Query, Row), Quizzes ).

getAllUserAnsweredQuizzes(UserId,Quizzes) :-
    format(atom(Query), "SELECT q.quiz_id, q.name, q.topic, ua.user_answer_id,
        ua.score, ua.rating, ua.created_at FROM quiz q, user_answer ua
        WHERE q.quiz_id = ua.quiz_id AND ua.user_id = '~w';",[UserId]),
    findall( Row, sqlite_query(db, Query, Row), Quizzes ).

getAllUserAnsweredQuizzesUnique(UserId,Quizzes) :-
    format(atom(Query), "SELECT DISTINCT q.quiz_id,q.name,q.topic,q.user_id,q.created_at
        FROM quiz q, user_answer ua WHERE q.quiz_id = ua.quiz_id AND
        ua.user_id = '~w';",[UserId]),
    findall( Row, sqlite_query(db, Query, Row), Quizzes ).

getAllQuizzesAnswers(UserId,Quizzes) :-
    format(atom(Query), "SELECT q.quiz_id, q.name, q.topic, ua.user_answer_id,
        ua.score, ua.rating, ua.created_at FROM quiz q, user_answer ua
        WHERE q.quiz_id = ua.quiz_id AND ua.user_id = '~w';",[UserId]),
    findall( Row, sqlite_query(db, Query, Row), Quizzes ).

updateQuiz(QuizId, NewName, NewTopic) :-
    format(atom(Query), "UPDATE quiz SET name = '~w', topic = '~w' WHERE quiz_id = '~w';",
        [NewName,NewTopic,QuizId]),
    sqlite_query(db, Query, _).

deleteQuiz(QuizId) :-
    format(atom(Query), "DELETE FROM quiz WHERE quiz_id = '~w';",[QuizId]),
    sqlite_query(db, Query, _).

printQuiz(Quiz) :- 
    Quiz = row(_,Name,Topic,_,_),
    format('Nome: ~w, Tópico: ~w\n',[Name,Topic]).

% formato da entrada: [row(QuizId,Name,Topic,UserId,CreatedAt)]
printQuizzes([],_).
printQuizzes([H|T],Index) :-
    H = row(_,Name,Topic,_,_),
    format('~d - Nome: ~w, Tópico: ~w\n',[Index,Name,Topic]),
    NIndex is Index + 1, 
    printQuizzes(T, NIndex).

printQuizAnswers([], _).
printQuizAnswers([H|T], Index) :-
    H = row(_,QuizName,QuizTopic,_,Score,Rating,CreatedAt),
    format('~d - Quiz: ~w, Tópico: ~w, Pontuação: ~2f, Avaliação: ~d, Respondeu em: ~w\n',
        [Index,QuizName,QuizTopic,Score,Rating,CreatedAt]),
    NIndex is Index + 1, 
    printQuizAnswers(T, NIndex).
